Add Vocos model

Manalelaidouni opened the pull request:
# What does this PR do?

This PR aims at integrating  `Vocos` model to `transformers`.

Vocos is a neural vocoder designed for high quality audio synthesis in TTS pipelines and related tasks, outpeforms `HifiGan` and it is significantly faster. It has 2 main variants : 

- `VocosModel` can be used as a standalone vocoder in audio generation pipeline, the goal is to use it as a drop in vocoder in YuE model. It can also be used together with `VocosFeatureExtractor` to synthesis audio from mel-spectrogram features.
- `VocosWithEncodecModel` :  integrates the EnCodec neural audio codec model into Vocos for end-to-end audio compression and reconstruction.

This is a continuation of integrating model components for the new YuE model (mention in #36784).

## Who can review?

Anyone in the community is free to review the PR once the tests have passed.
@ArthurZucker @eustlb @ylacombe

Manalelaidouni committed:
add working vocos

Manalelaidouni committed:
add working vocos

Manalelaidouni committed:
update vocos

Manalelaidouni committed:
update vocos

ArthurZucker commented on src/transformers/models/vocos/modeling_vocos.py:
```diff
@@ -0,0 +1,412 @@
+        return hidden_states
+
+
+class VocosISTFT(nn.Module):
+    """
+    Performs the Inverse Short Time Fourier Transform (ISTFT) on a STFT coefficients to reconstruct audio in the time domain.
+    It computes ISTFT differently depending on padding:
+        if `center` : uses PyTorch's built-in ISTFT implementation since it uses `center=True` by default.
+        if `same` : uses custom implementation of ISTFT with the overlap-add method, since the Pytorch version fails the
+        Nonzero Overlap Add (NOLA) condition when center is False. See issue: https://github.com/pytorch/pytorch/issues/62323
```
not entirely sure this makes sense to be in the modeling as it does not hold any weights!

Manalelaidouni replied:
You're right! I removed the ISTFT class.

ArthurZucker reviewed (commented):
Nice! My main comment is to remove the hidden states post processing!

Manalelaidouni committed:
refactor vocos head

Manalelaidouni committed:
refactor vocos head

Manalelaidouni committed:
fix docstring

Manalelaidouni committed:
fix docstring

Manalelaidouni committed:
nit

Manalelaidouni committed:
nit

Manalelaidouni committed:
Merge branch 'huggingface:main' into add-vocos-model

Manalelaidouni committed:
fix output mismatch

Manalelaidouni committed:
fix output mismatch

Manalelaidouni committed:
update checkpoint conversions

Manalelaidouni committed:
update checkpoint conversions

ArthurZucker commented on src/transformers/models/vocos/modeling_vocos.py:
```diff
@@ -0,0 +1,414 @@
+    def __init__(self, config: VocosConfig):
+        super().__init__()
+        self.out_proj = nn.Linear(config.hidden_dim, config.n_fft + 2)
+        if config.spec_padding not in ["center", "same"]:
+            raise ValueError("padding must be `center` or `same`")
```
typing in the config is enough for this

ArthurZucker commented on src/transformers/models/vocos/feature_extraction_vocos.py:
```diff
@@ -0,0 +1,253 @@
+            periodic=True,
+        )
+
+        self.mel_filters = mel_filter_bank(
```
nice!z

Manalelaidouni replied:
Thanks!

ArthurZucker reviewed (approved):
Sorry for my late review!

Manalelaidouni committed:
Merge branch 'main' into add-vocos-model

Manalelaidouni committed:
Merge branch 'add-vocos-model' of https://github.com/Manalelaidouni/transformers into add-vocos-model

Manalelaidouni committed:
fix adaptive layer norm

Manalelaidouni commented on src/transformers/models/vocos/modeling_vocos.py:
```diff
@@ -0,0 +1,414 @@
+        self.eps = config.layer_norm_eps
+        self.hidden_dim = config.hidden_dim
+        self.use_adaptive_norm = config.use_adaptive_norm
+        if self.use_adaptive_norm:
```
Yes, mel checkpoint uses standard LayerNorm and the Encodec one uses adaptive norm.

Manalelaidouni commented:
Thanks  @ArthurZucker all comments addressed!

Manalelaidouni commented:
Hey @Arthur! just checking in on this PR, I think we’re all set as the CI failures appear unrelated, is there anything else needed from my side? Thanks again!

Manalelaidouni committed:
fix conflict

Manalelaidouni committed:
fix auto

Manalelaidouni committed:
nit

Manalelaidouni commented:
Hey @ArthurZucker! CI is all green now should be good to go 👍

ebezzam commented on src/transformers/models/vocos/modeling_vocos.py:
```diff
@@ -0,0 +1,413 @@
+        self.head = VocosISTFTHead(config)
+        self.post_init()
+
+    def _audio_codes_to_features(self, codes: torch.LongTensor) -> torch.FloatTensor:
+        r"""
+        Transform a sequence of discrete tokens (codes) into feature embeddings using codebook weights.
+
+        Args:
+            codes (torch.LongTensor): Discrete codes of shape (n_codebooks, seq_len) or
+                (n_codebooks, batch_size, seq_len).
+
+        Returns:
+            torch.FloatTensor: Feature embeddings of shape (batch_size, hidden_dim, seq_len).
+        """
+        if codes.dim() == 2:
+            codes = codes.unsqueeze(1)
+        num_bins = self.encodec_model.quantizer.codebook_size
+        offsets = torch.arange(0, num_bins * codes.size(0), num_bins, device=codes.device).reshape(-1, 1, 1)
+        embeddings_idxs = codes + offsets
+        features = F.embedding(embeddings_idxs, self.codebook_weights).sum(dim=0)
+        return features.transpose(1, 2)
```
this function would also be in the VocosProcessor so that features are directly outputted by the processor that can be fed into the model.

And bandwidth can also be an outputs of the processor so that it can be used by the model.

ebezzam commented on src/transformers/models/vocos/modeling_vocos.py:
```diff
@@ -0,0 +1,413 @@
+        residual = hidden_states
+        hidden_states = self.dwconv(hidden_states)
+        hidden_states = hidden_states.transpose(1, 2)
+        if isinstance(self.norm, VocosAdaptiveLayerNorm):
```
why not store `self.use_adaptive_norm = config.use_adaptive_norm` in `__init__` and can then check this variable?

ebezzam commented on src/transformers/models/vocos/modeling_vocos.py:
```diff
@@ -0,0 +1,413 @@
+        audio: Optional[torch.FloatTensor] = None,
+        codes: Optional[torch.LongTensor] = None,
+        *,
+        bandwidth_id: torch.LongTensor,
```
and you can internally store a reverse mapping (from bandwidth to index) to use correct bias inside `VocosAdaptiveLayerNorm` when calling `hidden_states = self.backbone(hidden_states, bandwidth_id)` below

ebezzam commented on src/transformers/models/vocos/modeling_vocos.py:
```diff
@@ -0,0 +1,413 @@
+        hidden_states = self.backbone(hidden_states, bandwidth_id)
+        reconstructed_audio = self.head(hidden_states)
+
+        if return_codes:
+            return reconstructed_audio, codes
```
with processor can remove this, as the processor will return `codes`

ebezzam commented on src/transformers/models/vocos/modeling_vocos.py:
```diff
@@ -0,0 +1,413 @@
+        self,
+        audio: Optional[torch.FloatTensor] = None,
+        codes: Optional[torch.LongTensor] = None,
+        *,
```
what is this for?

ebezzam commented on tests/models/vocos/test_feature_extraction_vocos.py:
```diff
@@ -0,0 +1,241 @@
+# Copyright 2023 HuggingFace Inc.
```
update year

ebezzam commented on tests/models/vocos/test_modeling_vocos.py:
```diff
@@ -0,0 +1,630 @@
+        self.speech = ds[0]["audio"]["array"]
+
+    def test_inference(self):
+        EXPECTED = torch.tensor(
```
could you add link to reproducer for these expected outputs? Like [here](https://github.com/huggingface/transformers/blob/0a21e870c76262eb54cf6d7fde0f6c22b962204c/tests/models/xcodec/test_modeling_xcodec.py#L430). It will help for reproducibility and future debugging 🙂

ebezzam commented on utils/check_repo.py:
```diff
@@ -396,6 +396,7 @@
     "CsmDepthDecoderModel",  # Building part of a bigger model
     "CsmDepthDecoderForCausalLM",  # Building part of a bigger model
     "CsmForConditionalGeneration",  # Building part of a bigger model
+    "VocosModel",
```
We want to test this model, so you can remove from this list!

ebezzam reviewed (changes requested):
@Manalelaidouni great work, thanks for the PR 👏 

It may seem like a lot of comments, but it's mainly about being consistent with others parts of the repo.  

The main idea is to create a `VocosProcessor` inside `processing_vocos.py` that wraps the feature extractor and EnCodec into a single module. You can see [Dia](https://github.com/huggingface/transformers/blob/main/src/transformers/models/dia/processing_dia.py) as an example for audio.

You'll have to remove the requirement on a text tokenizer by overwriting the attributes (for example [here](https://github.com/huggingface/transformers/blob/0a21e870c76262eb54cf6d7fde0f6c22b962204c/src/transformers/models/siglip2/processing_siglip2.py#L62-L65) they remove the requirement on a feature extractor and add a required image processor):
```python
class VocosProcessor(ProcessorMixin):

    attributes = ["feature_extractor", "audio_tokenizer"]
```
This way we can move quite a bit of code away from `modeling_vocos.py` and avoid duplicating the pretrained weights of EnCodec.

And your neat example of using Bark tokens for TTS can also be added as a method to the processor!

`VocosModel` and `VocosWithEncodec` may end up looking very similar in the end! So we may even be able to simplify things after 🙂

ebezzam commented:
@Manalelaidouni you're welcome! It's much appreciated all the models you're adding and I'm happy to help 😀

And like YuE, other models use Vocos (xcodec2, Audio Flamingo 3) so this will be a great addition!

And no rush if you want to take a break after your project! All the best for that 💪

Manalelaidouni committed:
Merge branch 'huggingface:main' into add-vocos-model

Manalelaidouni committed:
add VocosProcessor and refactor

Manalelaidouni committed:
nit

Manalelaidouni committed:
clean up

ebezzam committed:
Update docs.

Manalelaidouni committed:
add bark example to docs

Manalelaidouni committed:
make fixture file shorter

Manalelaidouni commented on docs/source/en/model_doc/vocos.md:
```diff
@@ -0,0 +1,180 @@
+# generates EnCodec audio tokens from semantic tokens
+coarse_codes = generate_coarse(semantic_tokens, history_prompt=None, temp=0.7, silent=False,use_kv_caching=True)
+fine_codes = generate_fine(coarse_codes, history_prompt=None, temp=0.5)
+bark_codes = torch.from_numpy(fine_codes)
```
I added the example to docs;

ebezzam committed:
Update docs.

ebezzam committed:
Fix tensor device for tests.

ebezzam committed:
Nits

ebezzam reviewed (commented):
Thanks for the changes! As mentioned earlier, it's looking much nicer with using the processor 🙂 

I left a couple comments for batch processing + integration tests to be similar to others in the library. I imagine people may be interested in using Vocos on a batch of audio as well.

To that end, could you also add a batch integration test? (for example creating a batch like [this](https://github.com/huggingface/transformers/blob/main/tests/models/dac/test_modeling_dac.py#L841-L843) in your tests).

I also pushed a couple changes:
- Correcting device placement for the tests.
- In another PR, I was told to remove type hinting when the docs already make it clear. So I've done that in the configuration file.
- For `VocosAdaptiveLayerNorm`, since you raise an error if `cond_embedding_id` is not provided, I made it a mandatory input.
- Simplify the processor to have a single `__call__` method so it is clearer how inputs are processed (more of a Transformers way of doing things rather than breaking into smaller functions).

Manalelaidouni committed:
recreate fixtures

Manalelaidouni committed:
add torch backend  + batching support

Manalelaidouni committed:
add tests for both numpy and torch backends

Manalelaidouni committed:
add batch integration test for mel and encodec

Manalelaidouni committed:
make tests pass

Manalelaidouni committed:
update fixtures and tests

Manalelaidouni committed:
Merge branch 'main' into add-vocos-model

Manalelaidouni commented on tests/models/vocos/test_modeling_vocos.py:
```diff
@@ -0,0 +1,307 @@
+            "https://huggingface.co/datasets/hf-internal-testing/dummy-audio-samples/resolve/main/bcn_weather.mp3"
+        )
+        audio = librosa.load(BytesIO(urlopen(audio_url).read()), sr=24000)[0]
+        self.audio = torch.tensor(audio, dtype=torch.float32).unsqueeze(0)
```
done

Manalelaidouni commented:
Hey @ebezzam! thanks for the input I've addressed all the requested changes, mainly I've added the torch backend alongside the numpy one, added tests for both backends and implemented batch integration tests for both the mel and encodec variants.

ebezzam reviewed (commented):
Thanks @Manalelaidouni for the changes!

See my comments on the batch test. The test is meant to compare with expected output of the original model (like the single example test), rather than consistency between batched outputs.

Also sorry to point you to this Whisper example for computing mel features with torch. I think it might be better to use [MelSpectrogram](https://docs.pytorch.org/audio/2.7.0/generated/torchaudio.transforms.MelSpectrogram.html). I pointed to an example with Granite.

I also pushed a bit of cleaning to remove the `win_length` parameter as it is equal to `n_fft`.

(Side note: thanks for starting the YuE PR, I'll probably support you on that!)

Manalelaidouni committed:
add torchaudio backend + edit batching tests

Manalelaidouni committed:
nits

Manalelaidouni committed:
update test and  fixtures

Manalelaidouni committed:
Merge remote-tracking branch 'upstream/main' into add-vocos-model

Manalelaidouni committed:
cleanup

Manalelaidouni commented on tests/models/vocos/test_modeling_vocos.py:
```diff
@@ -0,0 +1,387 @@
+
+    @unittest.skip(reason="VocosModel only has one output format.")
+    def test_model_outputs_equivalence(self):
+        pass
```
done

Manalelaidouni commented on src/transformers/models/vocos/feature_extraction_vocos.py:
```diff
@@ -0,0 +1,288 @@
+
+        # always return batch
+        if not is_batched:
+            raw_speech = [raw_speech]
```
done

Manalelaidouni committed:
update feature extractor

Manalelaidouni commented on src/transformers/models/vocos/feature_extraction_vocos.py:
```diff
@@ -0,0 +1,294 @@
+                "Failing to do so can result in silent errors that might be hard to debug."
+            )
+
+        if is_torch_available() and isinstance(raw_speech, torch.Tensor):
```
Added this to address your todo comment in the processor to let feature extractor take tensor directly

Manalelaidouni commented:
hey @ebezzam I've switched the feature extractor to use torchaudio Melspectrogram instead of torch.stft for the mel feature computation, and updated the batch tests for both the mel and encodec variants as you requested, everything is now passing.

ebezzam reviewed (commented):
@Manalelaidouni just some minor changes to make best use of batch processing and other small things. Thank you!

Manalelaidouni committed:
Merge remote-tracking branch 'upstream/main' into add-vocos-model

ebezzam committed:
Fix small typo.

ebezzam committed:
Small fixes for passing integration tests.

ebezzam commented:
@Manalelaidouni I pushed some small fixes for the integration tests. One of them is still failing (`test_inference_batch_mel_vocos`), can you try the following command?
```bash
RUN_SLOW=1 pytest tests/models/vocos/test_modeling_vocos.py::VocosModelIntegrationTest
```

I'm getting the following (truncated) output:
```
tests/models/vocos/test_modeling_vocos.py::VocosModelIntegrationTest::test_batch_encodec_vocos PASSED                                                                                                          [ 25%]
tests/models/vocos/test_modeling_vocos.py::VocosModelIntegrationTest::test_inference_batch_mel_vocos FAILED                                                                                                    [ 50%]
tests/models/vocos/test_modeling_vocos.py::VocosModelIntegrationTest::test_inference_encodec_vocos PASSED                                                                                                      [ 75%]
tests/models/vocos/test_modeling_vocos.py::VocosModelIntegrationTest::test_inference_mel_vocos PASSED                                                                                                          [100%]

====================================================================================================== FAILURES ======================================================================================================
______________________________________________________________________________ VocosModelIntegrationTest.test_inference_batch_mel_vocos ______________________________________________________________________________

self = <tests.models.vocos.test_modeling_vocos.VocosModelIntegrationTest testMethod=test_inference_batch_mel_vocos>

    def test_inference_batch_mel_vocos(self):
        repo_id = self.mel_batch_expected["hf_repo_id"]
        processor = VocosProcessor.from_pretrained(repo_id)
        model = VocosModel.from_pretrained(repo_id).to(torch_device).eval()
    
        audios = self._load_datasamples(3)
    
        inputs = processor(audio=audios, return_tensors="pt").to(torch_device)
        hf_batch_output = model(**inputs).audio
    
        for i, saved in enumerate(self.mel_batch_expected["reconstructed_audio"]):
            expected = torch.tensor(saved, dtype=torch.float32, device=torch_device)
>           torch.testing.assert_close(
                hf_batch_output[i, : expected.shape[0]],
                expected,
                rtol=1e-4,
                atol=1e-4,
            )
E           AssertionError: Tensor-likes are not close!
E           
E           Mismatched elements: 20 / 20 (100.0%)
E           Greatest absolute difference: 0.0004398916498757899 at index (0,) (up to 0.0001 allowed)
E           Greatest relative difference: 1.1968653202056885 at index (0,) (up to 0.0001 allowed)

tests/models/vocos/test_modeling_vocos.py:333: AssertionError
```

ebezzam committed:
More Transformers compatible version, with naming and more amenable to modular.

ebezzam commented:
And also made some changes more amenable to Transformers, with naming and going back to original Vocos structure which will works well with Modular for Xcodec2 that I'm working on! #37868

Manalelaidouni committed:
refactor processor

Manalelaidouni committed:
update feature extractor torchaudio + spectogram_batch

Manalelaidouni committed:
nits

Manalelaidouni commented on src/transformers/models/vocos/feature_extraction_vocos.py:
```diff
@@ -0,0 +1,294 @@
+        if is_torchaudio_available():
+            input_features = [self._torch_extract_fbank_features(speech) for speech in raw_speech]
+        else:
+            input_features = [self._np_extract_fbank_features(speech) for speech in raw_speech]
```
I’ve updated both, made `_torch_extract_fbank_features()` take a batch directly and used `spectrogram_batch()` for the numpy version.

Manalelaidouni committed:
edit skipped tests reason

Manalelaidouni commented on tests/models/vocos/test_modeling_vocos.py:
```diff
@@ -0,0 +1,380 @@
+    def test_training_gradient_checkpointing(self):
+        pass
+
+    @unittest.skip(
+        reason="This architecture seem to not compute gradients properly when using GC, check: https://github.com/huggingface/transformers/pull/27124"
+    )
+    def test_training_gradient_checkpointing_use_reentrant(self):
+        pass
+
+    @unittest.skip(
+        reason="This architecture seem to not compute gradients properly when using GC, check: https://github.com/huggingface/transformers/pull/27124"
+    )
+    def test_training_gradient_checkpointing_use_reentrant_false(self):
+        pass
```
done!

Manalelaidouni committed:
Merge branch 'main' into add-vocos-model

ebezzam committed:
Nits.

ebezzam commented:
run-slow: vocos

ebezzam commented:
@Manalelaidouni Thanks for checking the tests. `test_inference_batch_mel_vocos` is still failing for me locally, and also on the GitHub Actions (see [here](https://github.com/huggingface/transformers/actions/runs/18277033466/job/52031520604)).

I'm guessing you obtained the expected outputs and ran the tests on CPU? When I run the integration tests on CPU (like below), they pass.
```
CUDA_VISIBLE_DEVICES="" RUN_SLOW=1 pytest tests/models/vocos/test_modeling_vocos.py::VocosModelIntegrationTest
```
Could you re-compute the expected outputs (at least for `test_inference_batch_mel_vocos`) on GPU?

ebezzam committed:
Updated expected outputs.

ebezzam commented:
@Manalelaidouni Thanks keep me posted! What is interested is that the slow test [passes](https://github.com/huggingface/transformers/actions/runs/18284649874/job/52056561043) (namely `test_inference_batch_mel_vocos` computed on GPU) with the new expected outputs, while `test_inference_batch_mel_vocos` now fails in [this test](https://app.circleci.com/pipelines/github/huggingface/transformers/148902/workflows/a64173c0-e784-4a6d-9803-fbeaa13b3991/jobs/1963519?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-checks-link&utm_content=summary).

From what I've seen `(Original-CPU, Transformers-CPU)` and `(Original-GPU, Transformers-GPU)` are close. What was being tested before, namely (Original-CPU, Transformers-GPU), was failing before. Maybe this new failing test is computed on CPU (I'll ask), and we may have to store expected outputs for both CPU and GPU.

ebezzam committed:
Add slow decorator.

ebezzam committed:
Fix import.

ebezzam committed:
Format.

ebezzam committed:
Move slow decorators to methods.

ebezzam committed:
Simplify feature extraction and more intuitive names.

Manalelaidouni commented:
> @Manalelaidouni Thanks keep me posted! What is interested is that the slow test [passes](https://github.com/huggingface/transformers/actions/runs/18284649874/job/52056561043) (namely `test_inference_batch_mel_vocos` computed on GPU) with the new expected outputs, while `test_inference_batch_mel_vocos` now fails in [this test](https://app.circleci.com/pipelines/github/huggingface/transformers/148902/workflows/a64173c0-e784-4a6d-9803-fbeaa13b3991/jobs/1963519?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-checks-link&utm_content=summary).
> 
> From what I've seen `(Original-CPU, Transformers-CPU)` and `(Original-GPU, Transformers-GPU)` are close. What was being tested before, namely (Original-CPU, Transformers-GPU), was failing before. ~Maybe this new failing test is computed on CPU (I'll ask), and we may have to store expected outputs for both CPU and GPU.~ Yes those are computed on CPU!
> 
> EDIT: just noticed, the `@slow` decorator was missing around the integration tests. I have added it.

@ebezzam I think I found why the same test passes in the one CI job  [here](https://github.com/huggingface/transformers/actions/runs/18284649874/job/52056561043)  and not [here](https://app.circleci.com/pipelines/github/huggingface/transformers/148902/workflows/a64173c0-e784-4a6d-9803-fbeaa13b3991/jobs/1963519?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-checks-link&utm_content=summary) , if you check the torchaudio wheels installed in each env, the first installs 2.8.0+cu126 and the second 2.8.0+cpu, the wheel mismatch makes torchaudio.transforms.MelSpectrogram produce slightly different numbers, i reproduced this in this gist https://gist.github.com/Manalelaidouni/d015f5b33f69fc1db5a1b1343e268fdb

Manalelaidouni committed:
Merge branch 'main' into add-vocos-model

Manalelaidouni committed:
make original vs hf feature extractor match on gpu

Manalelaidouni commented:
Hey @ebezzam, I just pushed an update the HF and original feature extractors now match on gpu and `torch.testing.assert_close(orig_mels, hf_mels, rtol=1e-3, atol=1e-3)` in the gist passes for both single and batched inputs, can you run the gpu tests in github actions when you get a chance?

ebezzam commented:
run-slow: vocos

ebezzam committed:
Merge branch 'add-vocos-model' of github.com:Manalelaidouni/transformers into add-vocos-model

ebezzam committed:
Simplify to just torch support.

ebezzam committed:
Standardize model inputs.

ebezzam committed:
Update docs

ebezzam committed:
Merge branch 'main' into add-vocos-model

ebezzam commented on src/transformers/models/vocos/feature_extraction_vocos.py:
```diff
@@ -0,0 +1,237 @@
+        sampling_rate: Optional[int] = None,
+        return_tensors: Optional[Union[str, TensorType]] = None,
+        return_audio_only: Optional[bool] = False,
+        device: Optional[str] = None,
```
With `device`, I did something similar to [GraniteSpeech](https://github.com/huggingface/transformers/blob/88e946e06233b6e470f44443b71ef45829ef5d30/src/transformers/models/granite_speech/feature_extraction_granite_speech.py#L69) except that I default to device of input audio. I feel like that is something user would expect? Namely when they place on a device that feature extractor would yield processed input on the same device?

ebezzam commented on src/transformers/models/vocos/modeling_vocos.py:
```diff
@@ -0,0 +1,368 @@
+        hidden_states = self.act(hidden_states)
+        hidden_states = self.pwconv2(hidden_states)
+        if self.layer_scale_parameter is not None:
+            hidden_states = self.layer_scale_parameter.to(hidden_states.device) * hidden_states
+        hidden_states = hidden_states.transpose(1, 2)
+        hidden_states = residual.to(hidden_states.device) + hidden_states
```
Needed to `.to(device)` for  `device_map="auto"` to work when loading form pretrained (VocosModel.from_pretrained(model_id, device_map="auto"))

ebezzam commented on src/transformers/models/vocos/modeling_vocos.py:
```diff
@@ -0,0 +1,368 @@
+    @auto_docstring
+    def forward(
+        self,
+        audio_spectrogram: Optional[torch.FloatTensor] = None,
+        input_features: Optional[torch.FloatTensor] = None,
```
To two types depending on the model:
- mel-variant: https://huggingface.co/hf-audio/vocos-mel-24khz
- encodec embeddings: https://huggingface.co/hf-audio/vocos-encodec-24khz

ebezzam commented on src/transformers/models/vocos/feature_extraction_vocos.py:
```diff
@@ -0,0 +1,237 @@
+        truncation: Optional[bool] = False,
+        sampling_rate: Optional[int] = None,
+        return_tensors: Optional[Union[str, TensorType]] = None,
+        return_audio_only: Optional[bool] = False,
```
for encodec variant, `return_audio_only` is used to cut processing before feature extraction (essentially to use input preperation and padding from the feature extractor)

ebezzam commented on src/transformers/models/vocos/processing_vocos.py:
```diff
@@ -0,0 +1,167 @@
+        if audio is not None:
+            if bandwidth is not None:
+                # pad audio into batch
+                fe_outputs = self.feature_extractor(audio, return_audio_only=True, device=device, **audio_kwargs)
```
where `return_audio_only` is used

ebezzam reviewed (commented):
@eustlb Vocos is nearly ready! I did a self-review to point out things for your feedback.

And here are checkpoints on the hub:
- Mel variant: https://huggingface.co/hf-audio/vocos-mel-24khz
- Encodec variant: https://huggingface.co/hf-audio/vocos-encodec-24khz

ebezzam commented:
run-slow: vocos

ebezzam commented:
@Manalelaidouni yes I wanted to update you on points that may help you for the YuE model addition.

And sorry for feedback that may seem "unnecessary" in hindsight. We're in the middle of a big standardization to make audio model interfaces/additions more intuitive. So rest assured, our iterations did help us better understand certain pain points and improvements. Thank you for your patience 🙏 

Some things you can keep in mind for your YuE model addition:
- Feature extractor should return a `padding_mask` so that we can get equivalent outputs between single and batch outputs (by using `padding_mask` to ignore/truncate corresponding samples). This was also why I moved the padding for encodec to the feature extractor (to have `padding_mask` for its audio inputs).
- Torch+batch processing should be default behavior for the feature extractor (for efficiency). Moreover, numpy is not necessary, which should hopefully simplify model addition and maintenance.*
- Using more intuitive variable names for the model inputs, e.g. `audio` and `audio_spectrogram` rather than `features`, `input_features`, etc... **

Hope that helps! And thanks again for your contributions 🙂 

_* For example, with Vocos we saw differences between torchaudio on CPU and GPU, and these would likely be present for numpy or torch implementations. Also, more changes would have been necessary for the latest numpy branch to make sure it was equivalent to the torch processing: move padding to before computing the spectrogram (also so "padding_mask" is equivalent), adding tests because none of them use an environment that doesn't have torch to make sure feature extraction is similar, etc_
_** I still kept `input_features` for this model because embeddings from the audio_codes could be passed an input, and we haven't thought of a standardized name for such inputs (also not very common)_

Manalelaidouni committed:
Merge branch 'main' into add-vocos-model

Manalelaidouni committed:
clean up

Manalelaidouni committed:
Merge branch 'main' into add-vocos-model

ebezzam committed:
Add gpu decorator.

Manalelaidouni committed:
undo warning

Manalelaidouni committed:
nits

ebezzam committed:
Add pad_to_multiple_of and use corresponding hop_length.

Manalelaidouni committed:
Merge branch 'main' into add-vocos-model

eustlb reviewed (commented):
Good work! Love how simple this looks, few questions on the modeling but close from merging 🤗

Manalelaidouni committed:
pad only batch

Manalelaidouni committed:
reproduce fixtures

Manalelaidouni committed:
minor correction

ebezzam committed:
Address comments.

ebezzam committed:
Reintroduce slow/gpu decorators.

Manalelaidouni committed:
change to old fixtures

ebezzam committed:
New istft utils and nits.

eustlb commented on docs/source/en/model_doc/vocos.md:
```diff
@@ -0,0 +1,207 @@
+
+
+The original code can be found [here](https://github.com/gemelo-ai/vocos) and original checkpoints [here](https://huggingface.co/charactr/vocos-mel-24khz) and [here](https://huggingface.co/charactr/vocos-encodec-24khz).
+
```
We'd want (best before merging if authors are responsive) to merge converted checkpoint directly to their repo, way better than having them under hf-audio

eustlb commented on docs/source/en/model_doc/vocos.md:
```diff
@@ -0,0 +1,207 @@
+    semantic_generation_config=semantic_generation_config,
+    coarse_generation_config=coarse_generation_config,
+    fine_generation_config=fine_generation_config,
+    codebook_size=bark.generation_config.codebook_size)
```
```suggestion
semantic_tokens = bark.semantic.generate(
    **bark_inputs,
    semantic_generation_config=semantic_generation_config,
)

coarse_tokens = bark.coarse_acoustics.generate(
    semantic_tokens,
    semantic_generation_config=semantic_generation_config,
    coarse_generation_config=coarse_generation_config,
    codebook_size=bark.generation_config.codebook_size,
)

fine_tokens = bark.fine_acoustics.generate(
    coarse_tokens,
    semantic_generation_config=semantic_generation_config,
    coarse_generation_config=coarse_generation_config,
    fine_generation_config=fine_generation_config,
    codebook_size=bark.generation_config.codebook_size,
)
```

eustlb commented on src/transformers/audio_utils.py:
```diff
@@ -609,6 +614,85 @@ def window_function(
+    else:
+        raise ValueError(f"Unsupported padding mode: {padding}. Supported modes are 'center' and 'same'.")
+
+    return audio
```
nice initiative but let's revert and simply add a TODO in the original code. Such an important function would require proper testing! work that has to be done when refacto audio_utils

eustlb commented on src/transformers/models/vocos/configuration_vocos.py:
```diff
@@ -0,0 +1,111 @@
+        self.hop_length = hop_length
+        self.istft_padding = istft_padding
+        self.bandwidths = list(bandwidths)
+        self.sampling_rate = sampling_rate
```
I do not see it used in the modeling code!

eustlb commented on src/transformers/models/vocos/modeling_vocos.py:
```diff
@@ -0,0 +1,302 @@
+        super().__init__(config)
+        self.backbone = VocosBackbone(config)
+        self.head = VocosISTFTHead(config)
+        self._bandwidth_to_id = {bandwidth: id for id, bandwidth in enumerate(config.bandwidths)}
```
There’s no reason to split the model into backbone and head. It doesn’t improve clarity, and the code is minimal in this case, so let’s just put everything in VocoModel.
This way, reading the forward method is enough to understand what’s happening.
`hidden_states = self.backbone(input_features, bandwidth_id)`
doesn’t convey any meaningful information. The reader still has to look up what the backbone does.
Similarly, calling a separate head misleads the reader into assuming it’s a simple linear layer (as is common), while in fact it’s linear + iSTFT.
Keeping everything in one place makes the model straightforward and transparent :)

eustlb commented on src/transformers/models/vocos/modeling_vocos.py:
```diff
@@ -0,0 +1,302 @@
+    def forward(
+        self,
+        audio_spectrogram: Optional[torch.FloatTensor] = None,
+        input_features: Optional[torch.FloatTensor] = None,
```
and what if the user provides both? we silently use only audio_spectrogram. This happens because there is no reason to provide both, so such a use case has no reason to be integrated in the model forward signature. `input_feature` and `audio_spectrogram` are both `input_features` (without considering the renaming, let's keep it to input_features for now).

eustlb commented on src/transformers/models/vocos/modeling_vocos.py:
```diff
@@ -0,0 +1,302 @@
+                config.layer_scale_init_value * torch.ones(config.hidden_size), requires_grad=True
+            )
+        else:
+            self.layer_scale_parameter = None
```
None of the two provided checkpoints have `layer_scale_init_value` set. There is no reason to set it to do this strange silent handling when < 0. I would even remove it and have it hardcoded as there is no reasons to have it as a config param.

eustlb commented on src/transformers/models/vocos/modeling_vocos.py:
```diff
@@ -0,0 +1,302 @@
+        else:
+            self.norm = nn.LayerNorm(config.hidden_size, eps=config.layer_norm_eps)
+        self.pwconv1 = nn.Linear(config.hidden_size, config.intermediate_size)
+        self.act = nn.GELU()
```
Nit but let's use Llama convention.
```suggestion
        self.act_fn = ACT2FN["gelu"]
```

eustlb reviewed (commented):
Here we see that different codepaths in the processor/ model are used in different situations, and cannot be used in a crossed manner. 
- if we use the mel spectrogram inputs, then we should not have an adaptative layernorm
- if we use the non-mel spectrogram inputs, then we should have an adaptative layernorm.

We should therefore have two models `Vocos`, and `VocosEncodec`.
`VocosEncodec` should be defined using modular, simply replacing the norm by a `VocosAdaptativeLayerNorm`. 

We should therefore have two models `Vocos`, and `VocosEncodec`.
`VocosEncodec` should be defined using modular, simply replacing the norm by a `VocosAdaptativeLayerNorm`.

Moreover it should have an embedding layer that does the input_features preparation that is currently in the processing for audio codes →  `inputs_embeds`, with a `torch.no_grad()`. I do get that we duplicate the weights necessary for codebook embedding in processor and the model, yet it's only 4MB and I'd rather fix the inputs of the model in token ids rather, so that it can be used without the processor given you have encodec codebook token ids. Morever, by passing codes directly, we do not need to pass the bandwidth id to VocosEncodec, which makes no sense to pass in the forward since it can be directly inferred from the `input_ids` shape.

`VocosEncodec` would simply have in it's modular something like:
```python
class VocosEncodecModel(VocosModel)
	def __init__(self, config):
		super.__init__(config)
		self.embed_tokens = nn.embedding()
		self.offsets = register_buffer(torch.arange(config.num_codebooks) * codebook_size, persistent=False)
		self.norms = nn.ModuleList(...)
        del self.norm

	def forward(
		input_ids, # shape (batch_size, seq_len, num_codebooks)
		inputs_embeds=None,
	)
		if (input_ids is None) ^ (inputs_embeds is not None):
            raise ValueError("You must specify exactly one of input_ids or inputs_embeds")
            
        num_codebooks = input_ids.shape[2]
        if num_codebooks not in self.config.supported_num_codebooks:
	        raise ValueError(f"{num_codebooks}...{self.config.supported_num_codebooks}")
	    else:
		    norm_idx = self.config.supported_num_codebooks.index(num_codebooks)
        self.norm = self.norms[norm_idx]
			
		if inputs_embeds is None:
			inputs_embeds = self.embed_tokens(audio_input_ids + self.offsets[:num_codebooks])
			
		super().forward(input_features=inputs_embeds)
```

ebezzam committed:
Update docs/source/en/model_doc/vocos.md

Co-authored-by: eustlb <94853470+eustlb@users.noreply.github.com>

ebezzam committed:
Update src/transformers/models/vocos/configuration_vocos.py

Co-authored-by: eustlb <94853470+eustlb@users.noreply.github.com>

ebezzam committed:
Flatten backbone and nits.

ebezzam committed:
Update mel conversion for flattening.

ebezzam committed:
Cleaner mel vocos model.

ebezzam committed:
Base simpler Vocos with encodec.

ebezzam committed:
Update convert for encodec variant and integration tests.

ebezzam committed:
Nits

ebezzam committed:
Nits before modular.

ebezzam committed:
From modular.

ebezzam committed:
Revert to input_features.

ebezzam committed:
Processor only for encodec variant.

ebezzam committed:
Make style

ebezzam committed:
Add codebook weights to Encodec model.

ebezzam committed:
Update modular and nits.

ebezzam committed:
Make style

Manalelaidouni commented:
hey @ebezzam are you done with the changes so I can handle the rest?

Manalelaidouni committed:
Merge upstream/main into add-vocos-model

Manalelaidouni committed:
Merge branch 'add-vocos-model' of https://github.com/Manalelaidouni/transformers into add-vocos-model

Manalelaidouni committed:
update feature extractor

Manalelaidouni committed:
update vocos modeling

Manalelaidouni committed:
update vocos encodec modeling

Manalelaidouni committed:
correct config

Manalelaidouni committed:
update fixtures

Manalelaidouni committed:
update and add pad test to feature extractor

Manalelaidouni committed:
update model tests

Manalelaidouni committed:
add processor tests

Manalelaidouni committed:
update processor

Manalelaidouni committed:
nits

Manalelaidouni committed:
update docs

Manalelaidouni committed:
allow EncodecModel as audio_tokenizer

Manalelaidouni committed:
Merge remote-tracking branch 'upstream/main' into add-vocos-model

Manalelaidouni committed:
correct weight initialization

Manalelaidouni committed:
fix modeling + feature extractor tests

Manalelaidouni committed:
update modular

Manalelaidouni committed:
ruff styling

Manalelaidouni committed:
update modular 2

Manalelaidouni committed:
nits

Manalelaidouni committed:
update auto mapping + tests

Manalelaidouni committed:
add codebook_weights buffer to initialization

Manalelaidouni committed:
allow unused config attribute

Manalelaidouni committed:
Merge branch 'main' into add-vocos-model

Manalelaidouni committed:
skip training tests

Manalelaidouni committed:
update docs

Manalelaidouni committed:
add test decorators

Manalelaidouni committed:
Merge branch 'main' into add-vocos-model
